crates/ruff_linter/src/rules/flake8_pyi/rules/bad_generator_return_type.rs

```diff
@@ -0,0 +1,191 @@
+
+    if !semantic
+        .resolve_call_path(map_subscript(returns))
+        .is_some_and(|call_path| {
+            matches!(
+                (name, call_path.as_slice()),
+                (
+                    "__iter__",
+                    ["typing" | "typing_extensions", "Generator"]
+                        | ["collections", "abc", "Generator"]
+                ) | (
+                    "__aiter__",
+                    ["typing" | "typing_extensions", "AsyncGenerator"]
+                        | ["collections", "abc", "AsyncGenerator"]
+                )
+            )
+        })
+    {
+        return;
+    };
+
+    // `Generator`` allows three type parameters; `AsyncGenerator`` allows two.
+    // If type parameters are present,
+    // Check that all parameters except the first one are either `typing.Any` or `None`;
+    // if not, don't emit the diagnostic
+    if let ast::Expr::Subscript(ast::ExprSubscript { slice, .. }) = returns {
+        let ast::Expr::Tuple(ast::ExprTuple { elts, .. }) = slice.as_ref() else {
+            return;
+        };
+        if matches!(
+            (name, &elts[..]),
+            ("__iter__", [_, _, _]) | ("__aiter__", [_, _])
+        ) {
+            if !&elts[1..].iter().all(|elt| is_any_or_none(semantic, elt)) {
+                return;
+            }
+        } else {
+            return;
+        }
+    };
+
+    // For .py files (runtime Python!),
+    // only emit the lint if it's a simple __(a)iter__ implementation
+    // -- for more complex function bodies,
+    // it's more likely we'll be emitting a false positive here
+    if !checker.source_type.is_stub() {
+        let mut yield_encountered = false;
+        for stmt in body {
+            match stmt {
+                ast::Stmt::Pass(_) => continue,
+                ast::Stmt::Return(ast::StmtReturn { value, .. }) => {
+                    if let Some(ret_val) = value {
+                        if yield_encountered
+                            && !matches!(ret_val.as_ref(), ast::Expr::NoneLiteral(_))
+                        {
+                            return;
+                        }
+                    }
+                }
+                ast::Stmt::Expr(ast::StmtExpr { value, .. }) => match value.as_ref() {
+                    ast::Expr::StringLiteral(_) | ast::Expr::EllipsisLiteral(_) => continue,
+                    ast::Expr::Yield(_) | ast::Expr::YieldFrom(_) => {
+                        yield_encountered = true;
+                        continue;
+                    }
+                    _ => return,
+                },
+                _ => return,
+            }
+        }
+    };
+
+    checker.diagnostics.push(Diagnostic::new(
+        GeneratorReturnFromIterMethod {
+            better_return_type: better_return_type.to_string(),
+            method_name: name.to_string(),
+        },
+        stmt.identifier(),
+    ));
+}
+
+fn is_any_or_none(semantic: &SemanticModel, expr: &ast::Expr) -> bool {
```

Comment:
We typically put the expression as the first argument since it's the "thing the method is about", and the `semantic` as the second since it's sort of auxiliary data.